### PR TITLE
feat(dpp): immutable properties on mutable document types

### PR DESCRIPTION
**Pull request by QuantumExplorer**
## Issue being fixed or feature implemented

Until now a document type was all-or-nothing: `documentsMutable: true` let a replace rewrite every property, and `false` froze the whole document. Apps regularly need a middle ground, such as a post whose `author` must never change while the body stays editable, or a creation-time reference that must not be repointed. Contracts had no way to express that, so it could only be "enforced" client-side, which is not enforcement.

## What was done?

- New doctype-level keyword **`immutable`** (meta-schema v3, protocol version 14): an array of top-level property names that are frozen at creation on a mutable document type. It sits next to `required` and `transient`.
- `DocumentTypeV2` gains `immutable_fields`, exposed as `DocumentTypeV2Getters::immutable_fields()`. V0 and V1 document types return an empty set.
- Parser generation 3 reads the keyword. Under full validation it refuses an entry that is not a top-level property (a nested path gets a hint to list the containing object instead), a `$` system property, and any list on a type with `documentsMutable: false`. The non-validating path, which stored contracts take, records the list as declared, so these lints can be tightened later without making a committed contract unreadable. Only the array shape is enforced on both paths.
- Enforcement lives in the replace action's state validation v1 (PV14): the action's existing `changed_data_fields` (changed value, added property, removed property) is intersected with the type's immutable list, and a hit fails the replace with the new **`DocumentImmutablePropertyChangedError`** (state code 40128, `StateError` discriminant 109, frozen-discriminant test extended). No new state read. Transfers, price updates and purchases carry no property data and are unaffected.
- Contract update (`validate_update` v1): the list may grow but never shrink (`DocumentTypeUpdateError`). The schema compatibility differ v1 now strips the top-level `immutable` key like `indices` and `required`, so a change is judged in exactly one place instead of hard-erroring as an unsupported keyword.
- A second list, **`immutableAllowSetting`** (`immutable_fields_allow_setting` in Rust), names immutable properties a replace may still set while the stored document has no value for them; once present they are frozen like the rest. Every entry must also be in `immutable` (parser lint). To tell a first-time set from a change, the replace action now also carries `added_data_fields`, the subset of `changed_data_fields` the stored document lacked. On contract update this list may lose entries at any time (a tightening) but may only gain one for a property that becomes immutable in the same update, so nothing already frozen starts accepting a late set.
- wasm-dpp maps the new error. The book's Documents chapter gets a section on the feature.
- Review round 1: `Value::equal_underlying_data` now recurses into maps (member order ignored) and arrays with the same integer-width leniency, so an untouched immutable object that storage reordered or narrowed no longer counts as changed; and the parser refuses a property that is both `transient` and `immutable`, which would have left the type permanently uneditable.

JavaScript discovery of the declarations is the stacked #4817.

Design choices worth a look:

- **Top-level names only.** `changed_data_fields` compares top-level properties, so listing an object freezes it whole, nested values included. Dotted paths would need a per-path comparison against the stored document; straightforward to add later if wanted.
- **Frozen includes presence.** An optional immutable property cannot be added after creation or removed, unless it is listed under `immutableAllowSetting`, which permits exactly one transition: absent to present.
- **Grow-only on update.** Adding an entry invalidates no stored document; removing one would break the promise documents were created under. The keyword names follow `required` / `transient`; happy to rename (`immutableProperties`, `immutableSettable`, ...) if preferred.

## How Has This Been Tested?

- rs-dpp, 14 parser tests in `try_from_schema/v3/immutable_tests.rs`: both validation modes, every lint, the array shape, the stored-contract path, PV13 gating (inert without validation, rejected by meta-schema v2 with it), and the dispatcher at latest. Update validator v1: removal refused, addition accepted, reorder accepted. Compat differ v1: an `immutable`-only diff is ignored. `StateError` frozen-discriminant test extended to 109.
- rs-drive-abci, 8 end-to-end tests in `batch/tests/document/immutable.rs` through `process_raw_state_transitions`: a mutable-only replace succeeds and is stored; changing, adding or removing an immutable property is a `PaidConsensusError` carrying `DocumentImmutablePropertyChangedError` with the property name and leaves storage untouched; an immutable object is frozen whole but does not false-positive when unchanged; the first changed property in name order is reported; a contract update removing an entry is refused; a contract update adding an entry is accepted and enforced afterwards.
- `immutableAllowSetting`: 7 more parser tests (parses on both modes, empty by default, entry not in `immutable` refused, unknown entry refused, non-string entry refused on both modes, stored-contract path, PV13 gating), 3 more update-validator tests (drop accepted, allowance on an already-immutable property refused, newly immutable property with allowance accepted), the compat test extended, and 5 more end-to-end tests (set once then frozen against change and removal while the rest stays editable; set at creation stays frozen; the allowance does not leak to other immutable properties; contract update granting the allowance to an already-immutable property refused; contract update adding a newly immutable property with the allowance accepted and enforced).
- Neighbouring suites re-run green: rs-dpp `document_type` and consensus error tests, drive-abci `replacement_tests`, `required_since`, `index_only` and `data_contract_update`. `cargo clippy -p dpp -p drive-abci --tests` and `cargo check -p wasm-dpp` clean.

## Breaking Changes

None. Protocol version 14 is unreleased; the keyword is rejected by meta-schema v2 and ignored by earlier parser generations, and nothing changes for contracts that do not use it.

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [x] I have made corresponding changes to the documentation if needed

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Document types can mark selected top-level properties as immutable after document creation.
  * Optional immutable properties may be configured to allow a one-time initial value.
  * Immutable properties are supported for mutable document types under protocol version 14.

* **Bug Fixes**
  * Replacements that change, add, or remove protected properties are now rejected with a specific validation error.
  * Contract updates enforce safe changes to immutable-property rules.
  * Comparisons correctly ignore object member order and integer-width differences.

* **Documentation**
  * Added documentation describing immutable properties, validation rules, and update behavior.

### CODE REVIEW

**Comment by thepastaclaw** [comment]
🕓 Review not started yet because the new head is waiting for the 30-minute push debounce.

- [ ] **Request normal review** — click when the PR is ready for review.
- [ ] **Request priority review** — click to move this review to the front of the queue.

_Commit d6f3c1af. Normal review starts when eligible; priority review starts as soon as a slot is available._

File: packages/rs-drive-abci/src/execution/validation/state_transition/state_transitions/batch/action_validation/document/document_replace_transition_action/state_v1/mod.rs [hunk 1]
```diff
@@ -48,6 +52,41 @@ impl DocumentReplaceTransitionActionStateValidationV1 for DocumentReplaceTransit
...
+            .contract
+            .document_type_for_name(document_type_name)?;
+        let immutable_fields = document_type.immutable_fields();
+        if !immutable_fields.is_empty() {
+            let allow_setting = document_type.immutable_fields_allow_setting();
+            let added_fields = self.added_data_fields();
+            if let Some(property) = self.changed_data_fields().iter().find(|field| {
+                immutable_fields.contains(*field)
+                    && !(allow_setting.contains(*field) && added_fields.contains(*field))
+            }) {
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/validation/state_transition/state_transitions/batch/action_validation/document/document_replace_transition_action/state_v1/mod.rs:78]: **[P2] Compare nested immutable values independently of representation**

The change set uses `equal_underlying_data`, which falls back to ordinary equality for nested objects. Storage can change integer widths (`U64(7)` becomes `U8(7)`) and reorder object members by schema position. Consequently, replacing only the mutable body while preserving an immutable object produces a paid `DocumentImmutablePropertyChangedError`. Both cases reproduce end-to-end. This check needs a recursive comparison that ignores equivalent integer representations and object key order.

File: packages/rs-dpp/src/data_contract/document_type/class_methods/try_from_schema/common/mod.rs [hunk 2]
```diff
@@ -1971,6 +1971,136 @@ pub(super) fn parse_index_only_keyword(schema: &Value) -> Result<bool, ProtocolE
...
+                    "; nested paths are not accepted, list the top-level property that contains \
+                     it to freeze it whole"
+                } else {
+                    ""
+                };
+                return Err(structure_error(format!(
+                    "document type \"{name}\" lists \"{property}\" as immutable, but it is not a \
+                     top-level property of the document type{hint}"
+                )));
+            }
```
Comment (QuantumExplorer) [inline packages/rs-dpp/src/data_contract/document_type/class_methods/try_from_schema/common/mod.rs:2084]: **[P2] Reject transient properties in the immutable list**

The parser accepts a property listed in `required`, `transient`, and `immutable`. Creation discards its value, so every replacement supplying it fails the immutable check—even when unchanged—while omitting it fails required-field validation. This makes the accepted document type permanently uneditable, and the grow-only rule prevents removing the restriction afterward. Reject immutable/transient overlap during full schema validation. Both replacement failures reproduce end-to-end.

**Review by QuantumExplorer — COMMENTED** [review]
Two issues reproduced on this revision, detailed inline. Validation: 30 targeted DPP tests and all 13 existing immutable-property end-to-end tests passed; three temporary end-to-end regression cases reproduced the issues below. Original sources were restored.

File: packages/rs-drive-abci/src/execution/validation/state_transition/state_transitions/batch/action_validation/document/document_replace_transition_action/state_v1/mod.rs (hunk 1 above)
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/execution/validation/state_transition/state_transitions/batch/action_validation/document/document_replace_transition_action/state_v1/mod.rs:78]: Fixed in d6f3c1af20. `Value::equal_underlying_data` now recurses: maps compare member by member regardless of order, arrays position by position, both with the existing integer-width and bytes leniency, so an object storage reordered by position or narrowed (U64 7 to U8 7) reads as unchanged. Since the replace transformer builds `changed_data_fields` with it, the state check needed no change. Regression: the e2e object test now creates `meta` with members in reverse position order and a U64 integer, replaces only `body` (accepted), then changes the integer and the string (each refused). Seven platform-value unit tests cover order, width, nesting, arrays, and the negative cases.

File: packages/rs-dpp/src/data_contract/document_type/class_methods/try_from_schema/common/mod.rs (hunk 2 above)
Comment (QuantumExplorer) [reply packages/rs-dpp/src/data_contract/document_type/class_methods/try_from_schema/common/mod.rs:2084]: Fixed in d6f3c1af20. The parser refuses a property listed under both `transient` and `immutable` under full validation ("a transient property is never stored, so every replace that supplies it would be refused as changing an immutable property"). Unit test `rejects_a_transient_property`; meta-schema description and the book bullet updated.

File: packages/rs-platform-value/src/eq.rs [hunk 3]
```diff
@@ -165,8 +170,126 @@ impl Value {
...
+        match (self, other) {
+            (Value::Map(this), Value::Map(that)) => {
+                this.len() == that.len()
+                    && this.iter().all(|(key, value)| {
+                        that.iter()
+                            .find(|(other_key, _)| key.equal_underlying_data(other_key))
+                            .is_some_and(|(_, other_value)| {
+                                value.equal_underlying_data(other_value)
+                            })
+                    })
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-platform-value/src/eq.rs]: Fixed in c10eeb8b83: the Map branch now matches entries one-to-one, consuming each matched entry so a duplicated key cannot be satisfied twice by one entry on the other side. Unit test `maps_match_entries_one_to_one` covers it.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by QuantumExplorer** [comment]
Reviewed
